pinot-core/src/main/java/com/linkedin/pinot/core/common/DataFetcher.java

```diff
@@ -33,47 +35,49 @@
 public class DataFetcher {
   private static final BlockId BLOCK_ZERO = new BlockId(0);
 
-  private final Map<String, BaseOperator> _columnToDataSourceMap;
   private final Map<String, Dictionary> _columnToDictionaryMap;
   private final Map<String, BlockValSet> _columnToBlockValSetMap;
   private final Map<String, BlockValIterator> _columnToBlockValIteratorMap;
   private final Map<String, BlockMetadata> _columnToBlockMetadataMap;
 
+  // Map from MV column name to max number of entries for the column.
+  private final Map<String, Integer> _columnToMaxNumMultiValuesMap;
+
+  // Re-usable array for all dictionary ids in the block, of a single valued column
+  private final Map<String, int[]> _columnToDictIdsMap;
+
+  // Re-usable array to store MV dictionary id's for a given docId
+  private final Map<String, int[]> _columnToTempDictIdsMap;
```

Comment:
For an example, if clients try to fetch two blocks data, the data from the first block will be override. We should not rely on clients to make the correct order of data fetch, but let the clients pass in a buffer to fill the data.